### DISCUSSION TITLE
How to do locale fallback lookup?

### DISCUSSION THREAD

**Discussion by faassen** [unicode-org/icu4x#3955] [Q&A] [answered]
I think that when I ask, for instance, for a collator given a locale, and that specific locale does not exist, it cannot find a collator. Instead I'd like it to fall back on another locale if possible.

I also found this document:

https://docs.google.com/document/d/1Mp7EUyl-sFh_HZYgyeVwj88vJGpCBIWxzlCwGgLCDwM/edit#heading=h.3jl4998lrpbp

which seems to have inspired this code:

https://docs.rs/icu_provider_adapters/latest/icu_provider_adapters/fallback/index.html

I want to try using this, but I don't understand how to turn a `BlobDataProvider` into a `LocalFallbackProvider`, because it doesn't have a `try_new_with_buffer_provider` constructor.

Is this intentional? Is it that buffer providers don't provide enough information yet, and this is not yet in the CLDR, as hinted at here:

https://github.com/unicode-org/icu4x/issues/2243

But maybe I'm on the wrong tack altogether. I can implement "subtag stripping" (as referenced by the google doc) manually but I was hoping there was something already.

So let's backtrack to the base question: how do I implement a locale fallback mechanism using icu4x?

**Comment by Manishearth** [comment, marked as answer]
The constructors you need are available with the `serde` feature enabled, looks like for this crate we do not have the `all-features = true` docs.rs key set for 1.2 (it's fixed in 1.3), but the functions exist. A local docs build should turn up the functions you need.

**Reply by sffc** [reply]
> looks like for this crate we do not have the all-features = true docs.rs key set for 1.2 (it's fixed in 1.3)

oops!

**Reply by faassen** [reply]
Ah, thanks, I will add the feature!

**Comment by sffc** [comment]
I just hit merge on https://github.com/unicode-org/icu4x/pull/3334 which adds a tutorial example for this.

**Comment by faassen** [comment]
The missing serde dependency was indeed the problem. Now I'm using a fallback provider, and an either provider too (as fallback can be turned on and off), and it all appears to work the way I expect! Thank you so much for the good work!
